### PR TITLE
Add default implementations for FixedWidthInteger.dividingFullWidth

### LINKED ISSUE #54509: [SR-12073] dividingFullWidth is missing overflow checks
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-12073      |
   |Radar             | rdar://problem/58839693         |
   |Original Reporter | @stephentyrone      |
   |Type              | Bug    |

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Standard Library    |
   |Labels            | Bug        |
   |Assignee          | @stephentyrone      |
   |Priority          | Medium      |

   

   md5: c29bc2f189c69046b795880fe2c71338

  </details>

**Issue Description:**

Reported in the forums here: <https://forums.swift.org/t/signed-integer-big-division-wraps-around-to-negative/32817>

We should also take this opportunity to provide a generic implementation on FixedWidthInteger, like I did a few months ago for the full-width multiply.

### PR DESCRIPTION
**Pull request by stephentyrone**
These are provided for FixedWidthInteger & UnsignedInteger (the base implementation, following Knuth's Algorithm D) and SignedInteger (converting to magnitudes and calling the former). Previously no default implementations were available, requiring every type to implement these operations.

These defaults will not be optimal for large fixed-width integers, so types vending Int512 or similar integers should still provide their own implementations, but they are unconditionally available as a fallback, which simplifies the process of writing such types, and work well enough as a fallback for modest fixed-width integer types like Int64 or 32b or smaller platforms or Int128 on 64b platforms.

Additionally rework the concrete implementations to guarantee that we always trap when the quotient is not representable, and to improve performance for 64b integers on `arm64_32`

Fixes #54509.

### CODE REVIEW

**Comment by stephentyrone** [comment]
@swift-ci test

**Comment by stephentyrone** [comment]
@swift-ci test

File: stdlib/public/core/Integers.swift [hunk 1]
```diff
@@ -3491,9 +3502,124 @@ extension UnsignedInteger where Self: FixedWidthInteger {
...
+    // half-words (as here), that can never happen, because we use the full
+    // divisor in the check for the while loop.
+    func generateHalfDigit(high: Self, low: Self) -> Self {
+      // Get q̂ satisfying a = vh q̂ + r̂ with 0 ≤ r̂ < vh:
+      var (q̂, r̂) = high.quotientAndRemainder(dividingBy: vh)
+      // Knuth's "Theorem A" establishes that q̂ is an approximation to
+      // the quotient digit q, satisfying q ≤ q̂ ≤ q + 2. We adjust it
+      // downward as needed until we have the correct q.
+      while q̂ > mask || q̂ &* vl > (r̂ &<< n_2 | low) {
+        q̂ &-= 1
```
Comment (phausler) [inline stdlib/public/core/Integers.swift:3600]: ❤️ the use of circumflex

File: stdlib/public/core/Integers.swift [hunk 2]
```diff
@@ -3491,9 +3502,124 @@ extension UnsignedInteger where Self: FixedWidthInteger {
...
+    // high is the high word of the quotient for this sub-division.
+    // low is the low half-word of the quotient for this sub-division (the
+    //     high half of low must be zero).
+    //
+    // returns the quotient half-word digit. In a more general setting, this
+    // computed digit might be one too large, which has to be accounted for
+    // later on (see Knuth, Algorithm D), but when the divisor is only two
+    // half-words (as here), that can never happen, because we use the full
+    // divisor in the check for the while loop.
+    func generateHalfDigit(high: Self, low: Self) -> Self {
```
Comment (phausler) [inline stdlib/public/core/Integers.swift:3593]: I presume this as an inner function is 100% specialized and inlined

File: stdlib/public/core/Integers.swift [hunk 3]
```diff
@@ -3491,9 +3502,124 @@ extension UnsignedInteger where Self: FixedWidthInteger {
...
+    let v = self &<< lz
+    let uh = dividend.high &<< lz | low >> (Self.bitWidth - lz)
+    let ul = low &<< lz
+    
+    // Now we have a normalized dividend (uh:ul) and divisor (v). Split
+    // v into half-words (vh:vl) so that we can use the "normal" division
+    // on Self as a word / halfword -> halfword division get one halfword
+    // digit of the quotient at a time.
+    let n_2 = Self.bitWidth/2
+    let mask = Self(1) &<< n_2 &- 1
```
Comment (phausler) [inline stdlib/public/core/Integers.swift:3561]: depending on the register spill here; would it be worthwhile to optimize by adding additional variables like something to represent the masked value of `ul | mask` etc?

**Review by phausler — APPROVED** [review]
The algorithm looks spot on to me; I presume we already have tests that cover this?

**Comment by stephentyrone** [comment]
Most normal standard library tests won't exercise it because standard library integer types specialize the operation. I wrote a bunch of standalone tests for it, which I'll see about rolling into the test suite.

File: stdlib/public/core/Integers.swift (hunk 3 above)
Comment (stephentyrone) [reply stdlib/public/core/Integers.swift:3561]: It ought not to matter, so I'll hold off until we see a case where it does.

File: stdlib/public/core/Integers.swift (hunk 2 above)
Comment (stephentyrone) [reply stdlib/public/core/Integers.swift:3593]: So long as the outer function is specialized, yes.

**Review by glessard — APPROVED** [review]
Nice explanations along the way!

**Comment by stephentyrone** [comment]
@swift-ci test

**Comment by stephentyrone** [comment]
@swift-ci benchmark

**Comment by stephentyrone** [comment]
(We don't expect benchmarks to move much from this, but it's good to make sure.)

**Review by slavapestov — APPROVED** [review]
I love it!

**Comment by stephentyrone** [comment]
As expected, benchmarks are just noisy tests, nothing related to division. Importantly, no noticeable code size changes at all.

File: stdlib/public/core/Integers.swift (hunk 1 above)
Comment (natecook1000) [reply stdlib/public/core/Integers.swift:3600]: omg

**Pull request merged**